### PR TITLE
fix(fetch): preserve Markdown code boundaries and disclosure content

### PR DESCRIPTION
**Pull request by lanyue-llk**
## Why this change is needed

The tagged Moli 1.1.8 release still loses content structure in two HTML-to-Markdown paths, and optional UI stripping can remove product body text. These are output defects on frozen pages, independent of the WCXB word-overlap score.

### Preformatted code blocks — WCXB 5412

The input has a separate `ts` language label and `function` block. Release output joins them as `tsfunction`; this PR preserves the block boundary and the code lines. The same root cause appears on 5411, 5413 and 5423.

Frozen HTML, tagged Moli 1.1.8 output, and PR code-block output

### Product description — WCXB 4017, optional UI stripping

With `fetch --dump markdown --strip-mode ui`, release 1.1.8 drops the `<details>` subtree containing the product description. This PR keeps its body text while stripping UI chrome. This optional mode is **outside** the ordinary 504-page score. The [published report](https://artifact-site.lexmount.cn/v/7GfS87N6u4G1VWcEJlT9sG3j8YWV1Zjo) includes both outputs and their binary hashes.

Frozen description, tagged Moli 1.1.8 optional output, and PR optional output

### Adjacent inline code — WCXB 5458

The input has separate `<code>name</code><code>string</code>` nodes. Release 1.1.8 emits one `namestring` code value. The PR emits `` <code>name</code>`string` ``: CommonMark renders two adjacent `<code>` elements, so their semantic boundary survives without inventing text. The visible gap on this frozen page comes from its CSS `gap-2` class; a source-only HTML-to-Markdown converter has no computed layout and must not manufacture that space. The regression test checks the **rendered CommonMark HTML**, including empty code nodes and adjacent punctuation.

Frozen adjacent code, tagged Moli 1.1.8 output, and PR output

## Implementation and scope

- The preformatted traversal writes line breaks at structural block boundaries; phrasing elements stay inline, and explicit `<br>` remains a break.
- UI stripping no longer discards `details` and `summary` with their descendants.
- Adjacent code nodes use the inline HTML fallback only where separate Markdown spans would lose their boundary. HTML metacharacters use entities; other ASCII punctuation uses the complete CommonMark backslash-escape class.
- Image references are preserved. The plain-text evaluator counts image and link addresses, but removing valid references for a higher score would damage Markdown output.

The official Moli 1.1.8 binary and PR source tree were measured independently on the same 504 qualified frozen WCXB pages: both return nonempty content on 504/504; mean F1 is 52.5510% versus 52.5492%. Two pages rise, two fall, and 500 do not change. Mean recall changes +0.0067 percentage points, mean precision −0.0048, and mean page-level F1 −0.0018. Aggregate F1 is the mean of 504 page-level F1 values, not a recomputation from the headline mean precision and recall. On page 5458, reference recall rises from 98.45% to 100%, while the raw scorer also tokenizes the Markdown source and counts the inline HTML `code` tag names; that lowers precision and F1 even though the rendered output preserves the missing node boundary. The full responses, scorer results, binary hashes, optional-mode receipt, and same-window latency calls are in the [published WCXB report](https://artifact-site.lexmount.cn/v/7GfS87N6u4G1VWcEJlT9sG3j8YWV1Zjo). Release 1.1.8 and this PR have different version histories, so the full-cohort difference is a comparison of the two binaries rather than an isolated estimate of only these edits.

The 504-page results and figures are measured evidence for `384d1d2a`. Revision `4065e86c` generalizes punctuation escaping and expands compatibility scenarios; these numbers are not a new measurement of that revision. CI: https://github.com/lexmount/moli/actions/runs/35485830577.

Compatibility coverage retains text-row scrolling with explicit row heights, checks that UI stripping removes navigation and buttons while retaining open and closed disclosure content, and compares source HTML code values with the final Markdown rendering. The original adjacent-code fixture remains; its expected output changes because merging its two code nodes was the defect.

### CODE REVIEW

**Review by lanyue-llk — COMMENTED** [review]
Review summary: the PR explains its four target defects and includes regression tests; I found no mandatory PR template in the repository. Format, Clippy, and Nextest checks passed. The Runtime/CDP and deterministic Spider jobs were cancelled, so those contracts have no completed CI result for this head. The inline findings below are reproduced against this commit and concern content loss outside the cited fixtures. Please address them before merge.

File: moli-html2md/src/machine.rs [hunk 1]
```diff
@@ -428,6 +455,62 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
...
+    let encoded = parts.any(|part| part.eq_ignore_ascii_case("base64"));
+    let bytes = if encoded {
+        let Ok(bytes) = base64::engine::general_purpose::STANDARD.decode(payload) else {
+            return false;
+        };
+        bytes
+    } else {
+        percent_encoding::percent_decode_str(payload).collect::<Vec<_>>()
+    };
+    if kind.eq_ignore_ascii_case("image/gif") {
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs]: [P1] Can we avoid treating every 1×1 GIF/PNG as a noncontent placeholder? Dimensions alone do not determine visibility: a 1×1 colored image enlarged by CSS is real page content. I ran the converter with `Red status pixel`; this branch emits only `Red status pixel`, whereas the previous path emitted an image. The same applies to PNG. Please restrict removal to evidence of a placeholder, or retain ambiguous images so content is not lost.

File: moli-html2md/src/machine.rs [hunk 2]
```diff
@@ -428,6 +455,62 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
...
+            && bytes.get(16..24) == Some(&[0, 0, 0, 1, 0, 0, 0, 1][..]);
+    }
+    let Ok(svg) = std::str::from_utf8(&bytes) else {
+        return false;
+    };
+    let Ok(document) = roxmltree::Document::parse(svg) else {
+        return false;
+    };
+    let root = document.root_element();
+    root.tag_name().name() == "svg"
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs]: [P2] Can the SVG check account for visible root attributes? `` renders a colored 32×32 image, but this converter emits an empty string because the SVG has no child nodes. That contradicts the stated goal of retaining meaningful images. Checking only children is insufficient; visible paint/style on the root must prevent suppression.

File: moli-html2md/src/machine.rs [hunk 3]
```diff
@@ -234,7 +240,20 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
                     .attribute(node, "src")
                     .filter(|src| !src.is_empty())
                 {
-                    self.writer().image(alt, src, title);
+                    // Embedded placeholders carry no useful image. Keep
+                    // accessible alternate text without emitting a data URI.
+                    if is_noncontent_data_image(src) {
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs]: [P2] Can this decision consider `srcset`/`picture` before dropping the ``? For `Product`, browsers can select the real product image, but the converter now emits only `Product`. The old fallback at least preserved an image reference. Suppressing based solely on `src` loses image content on responsive pages; retain an image reference when a nonplaceholder candidate exists.

File: moli-html2md/src/machine.rs (hunk 1 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs]: Updated after further review: transparent or undecodable pixels do not prove an image has no page meaning. Commit 30c4f85b removes pixel-based image suppression entirely; every img with src keeps its Markdown image reference, including opaque and transparent 1×1 images. The regression test covers both.

File: moli-html2md/src/machine.rs (hunk 2 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs]: Updated after further review: 30c4f85b removes the SVG emptiness inference entirely. Both a childless SVG and one painted through root style retain their image references. This avoids deleting content based on a partial rendering model.

File: moli-html2md/src/machine.rs (hunk 3 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs]: Updated after further review: 30c4f85b retains img references regardless of src content, so srcset/picture can no longer cause the fallback image to be dropped. The converter does not guess a selected responsive candidate without layout/DPR context; existing fallback behavior remains.

**Review by lanyue-llk — COMMENTED** [review]
Deep re-review of `30c4f85b` against common ancestor `bffa3346`.

The previous three image findings are resolved: the final diff removes image suppression entirely and restores the dependency-free converter. Disclosure retention is implemented in the shared dump path, so it also applies to HTML/JSON dumps rather than only the reported Markdown case. The three remaining objectives and the adjacent-code behavior change are clearly described; I found no mandatory PR template.

I independently reran `cargo test -p moli-html2md`: all 290 existing tests passed. The current GitHub CI checks, including Format, workspace Clippy, Nextest, Runtime/CDP contracts and the deterministic Spider fixture, are successful. However, two additional content-preservation probes pass on the common-ancestor implementation and fail on this head. Both reproduced regressions are documented inline and should be fixed before merge.

A generality check also found that heading/list boundaries inside `<pre>` remain unhandled because the reused `is_block` helper excludes elements handled specially by the normal traversal. This is an incomplete part of the stated fix, not a newly introduced regression; it is included with the classifier finding so the repair can cover both false positives and false negatives.

Evidence question: could the 504-page WCXB result link its frozen-corpus/qualification manifest, exact candidate build revision, invocation and raw evaluator output? The description gives aggregate results and page numbers, but the PR currently does not provide enough evidence to independently reproduce those measurements.

File: moli-html2md/src/machine.rs [hunk 4]
```diff
@@ -380,6 +386,14 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
             // breaks without converting descendant formatting into Markdown.
             NodeKind::Element("br") if inline => self.raw.push('\n'),
             NodeKind::Document | NodeKind::Element(_) if depth + 1 < self.options.max_depth => {
+                if !inline
+                    && matches!(self.dom.node_kind(node), NodeKind::Element(tag) if is_block(tag))
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs]: [P2] Use a structural boundary classifier suitable for raw preformatted text

Can this avoid reusing the normal traversal's fallback `is_block` list? That list includes `output`, which is [phrasing content](https://html.spec.whatwg.org/multipage/form-elements.html#the-output-element). With `<pre>const answer = <output>42</output>;</pre>`, the common ancestor produces the code line `const answer = 42;`, while this head produces `const answer = \n42\n;` inside the fence. These added line breaks change preformatted content despite the stated promise to leave inline descendants in their original stream.

The same helper also omits `h1`–`h6`, `blockquote`, `ul`, `ol` and `li` because the normal traversal handles them in separate match arms. Consequently, the advertised block-boundary repair still emits `first()second()` for `<pre><ol><li>first()</li><li>second()</li></ol></pre>` and `tsfunction...` for `<pre><h3>ts</h3><code>function...</code></pre>`. Those latter omissions predate this PR; the newly added newline corruption does not. Please give raw traversal the appropriate structural classification and cover both inline descendants and the separately handled block tags.

File: moli-html2md/src/writer.rs [hunk 5]
```diff
@@ -171,6 +171,14 @@ impl<'a> Writer<'a> {
     }
 
     pub(crate) fn code_with_edges(&mut self, text: &str, preformatted: bool) {
+        // Each call represents a separate HTML code element. Keep adjacent
+        // elements distinct even when the source has no whitespace between them.
+        if self.code.is_some() {
+            self.flush_code();
+            if !text.is_empty() {
+                self.space = true;
+            }
```
Comment (lanyue-llk) [inline moli-html2md/src/writer.rs]: [P2] Preserve code separation across empty code elements

Can empty `<code>` elements be handled before flushing the previous code buffer, or can separation survive that flush? With `Options { preformatted_code: true, ..Default::default() }`, `<code>name</code><code></code><code>string</code>` now emits ``` `name``string` ```. Rendering that Markdown with the repository's `pulldown-cmark` helper yields `<p><code>name``string</code></p>`: two literal backticks have been introduced into the content. The common ancestor emits `` `namestring` `` without invented characters, so this is a regression in the public preformatted mode.

The empty element flushes `self.code` without setting `self.space`; when the following nonempty element arrives, `self.code` is already `None`, so this guard no longer inserts a separator. A comment-only code element triggers the same problem. The default mode already had the empty-element defect before this PR, so please cover both modes when fixing the shared boundary state instead of only testing two directly adjacent nonempty elements.

File: moli-html2md/src/machine.rs (hunk 4 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs]: Confirmed on 30c4f85b: inline <output> inserted line breaks, while heading/list descendants lacked structural breaks. Commit a6315618 separates raw structural boundaries from the normal fallback classifier and preserves explicit <br>. The scenario test renders the generated Markdown and checks the visible preformatted lines for <output>, heading, list, and <br>; it failed before the fix and passes after it. The complete moli-html2md suite passes.

File: moli-html2md/src/writer.rs (hunk 5 above)
Comment (lanyue-llk) [reply moli-html2md/src/writer.rs]: Confirmed in both default and preformatted modes on 30c4f85b. Commit a6315618 leaves the pending nonempty code value intact when an intervening <code> is empty, so the next visible code value gets the existing separator. The new scenario test covers empty and comment-only elements, parses Markdown back to HTML, and requires two distinct code nodes without invented backticks. It failed before and passes after the fix; the complete moli-html2md suite passes.

File: moli-html2md/src/machine.rs [hunk 6]
```diff
@@ -130,6 +131,11 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
                     self.writer().block(item.into(), separation, separation);
                 }
                 Task::RawChildren(Some(node), depth, inline) => self.raw_node(node, depth, inline),
+                Task::EndRawBlock => {
+                    if !self.raw.ends_with('\n') {
+                        self.raw.push('\n');
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs:136]: 这种判断确定是通用普适的吗？看起来认为加了换行符，我不确定

File: moli-html2md/src/machine.rs [hunk 7]
```diff
@@ -378,8 +384,16 @@ impl<'a, D: Dom + ?Sized> Machine<'a, D> {
...
             // Inline code normalizes line endings to spaces. Keep explicit
             // breaks without converting descendant formatting into Markdown.
-            NodeKind::Element("br") if inline => self.raw.push('\n'),
+            NodeKind::Element("br") => self.raw.push('\n'),
             NodeKind::Document | NodeKind::Element(_) if depth + 1 < self.options.max_depth => {
+                if !inline
+                    && matches!(self.dom.node_kind(node), NodeKind::Element(tag) if is_structural_block(tag))
+                {
+                    if !self.raw.is_empty() && !self.raw.ends_with('\n') {
+                        self.raw.push('\n');
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs:393]: 同上，确定这是最合适的修改方式吗

File: moli-html2md/src/machine.rs (hunk 6 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs:136]: [P2] 这里确实有可以复现的反例：空块会凭空制造代码的首行空白。

在 `ab749f63` 上输入 `<pre><div><!-- annotation --></div>print(42)</pre>`，生成的 fenced code 内容是 `"\nprint(42)\n"`；合并基线 `bffa3346` 是 `"print(42)\n"`。将 Markdown 再解析成 HTML 后仍有这条额外空行，所以这是内容变化，不只是 fence 的排版差异。空 `<div></div>` 和嵌套空块同样触发。

根因是空字符串也满足 `!self.raw.ends_with('\n')`：此时还没有任何文本，`EndRawBlock` 已经写入了分隔符。结构边界应分隔内容，而不应在第一段内容之前制造文本。最低限度应增加非空判断；我已验证该改动能让上述回归测试通过。修复时须保留源文本本来就有的首行换行，不能用最终 `trim` 处理。

因此，这段当前实现不够通用，但不能由此得出“所有块边界换行都应删除”：普通相邻非空块仍需要分隔。

File: moli-html2md/src/machine.rs (hunk 7 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs:393]: 需要区分结构化导出和视觉还原。`moli-html2md/README.md` 明确声明不计算 CSS、visibility 或 layout；因此，按结构标签给相邻块内容加分隔，是符合当前模块职责的策略，但它不能承诺在 `display:inline` 等样式覆盖下仍与视觉布局一致。没有必要为了这个 PR 把布局引擎耦合进转换器。

入口这段已有 `!self.raw.is_empty()`，目前确证的问题在出口：空块退出时仍可写入首个换行，具体复现见上一条评论的回复。建议统一入口/出口的边界语义，避免首尾空块制造内容，并分别覆盖空块、嵌套块、相邻非空块、显式 `<br>` 和源文本自带换行。原文换行与结构分隔应分别处理，不能一律增补，也不能一律删除。

这次实测支持的是上述具体修改；我没有把“人为补了换行”本身作为缺陷判据。

File: moli-html2md/src/writer.rs [hunk 8]
```diff
@@ -171,6 +171,19 @@ impl<'a> Writer<'a> {
...
+        // the next visible text or element decides whether it needs a gap.
+        if text.is_empty() {
+            return;
+        }
+        // Each call represents a separate HTML code element. Keep adjacent
+        // elements distinct even when the source has no whitespace between them.
+        if self.code.is_some() {
+            self.flush_code();
+            if !text.is_empty() {
+                self.space = true;
```
Comment (lanyue-llk) [inline moli-html2md/src/writer.rs]: [P2] Preserve source whitespace when separating code elements

Can code-node boundaries be retained without inserting visible text? With `preformatted_code: true`, `<code>git </code><code>status</code>` now becomes `` `git ` `status` ``. Parsing this with the repository's Markdown renderer yields `<p><code>git </code> <code>status</code></p>`: the original single space is now two spaces (`"git  status"`). The common ancestor preserves `"git status"`. This contradicts the public option's whitespace-preservation contract and is missed by the changed `inline_code_breaks_keep_separation_at_code_span_edges` expectation, which accepts an additional external space without asserting the concatenated rendered text.

The same unconditional separator changes `<code>--</code><code>help</code>` into `"-- help"` under the default mode. Separate code elements do not establish that the source contains an intervening space. A lossless Markdown/inline-HTML boundary can retain two code nodes and their exact text; I verified this with the same renderer. Please test both node separation and the complete rendered text, including trailing/leading spaces, rather than satisfying the former by changing the latter.

**Review by lanyue-llk — COMMENTED** [review]
Deep re-review of `ab749f63`, including the two human comments about synthesized line breaks.

The previous empty-code/backtick and inline-output classification findings are fixed, and all 292 existing `moli-html2md` tests pass locally. However, three additional content-preservation checks pass against common ancestor `bffa3346` and fail on this head. They identify two remaining problems: [an empty initial block invents a leading blank line](https://github.com/lexmount/moli/pull/643#discussion_r4045720943), and adjacent-code separation invents visible whitespace (inline finding below). Please fix these before merge.

The human comments have been answered with the reproduced case and the module's documented scope. Structural block separators are appropriate for this CSS-independent dump; claiming universal visual equivalence would not be appropriate. A lossless inline-HTML boundary was also tested: it preserves two code nodes and the original text without an added space.

The new viewport-relative scroll fixtures retain the existing observable scroll assertions; I found no defect in that fixture adjustment. The supplied WCXB manifest/profile links are accessible and bind the 504-page measurement to `30c4f85b`, and the description correctly states that the current head was not remeasured. At review time Format and Clippy are successful; Nextest and release builds are still pending, so the current head does not yet have a complete CI result.

Description-only suggestion: condense the review chronology and temporary scroll-test troubleshooting into final behavior and validation scope, keeping the pinned reproduction links.

File: moli-html2md/src/machine.rs (hunk 6 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs:136]: 已核对并修正。块边界换行只应分隔已有内容；空块退出时原实现会在第一段真实文本前凭空加一行。384d1d2a 在写入出口换行前检查已有内容，覆盖空块、嵌套空块、非空相邻块及源文本自带换行的场景测试。结构性分隔仍保留，因为相邻非空块需要它。

File: moli-html2md/src/machine.rs (hunk 7 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs:393]: 这处入口的结构分隔保留。转换器按 HTML 结构导出代码文本，不计算 CSS 布局；相邻非空块必须有边界，内联 output 则不应被拆行。之前的问题在空块出口凭空加换行，384d1d2a 已修正并通过转换器测试。

File: moli-html2md/src/writer.rs (hunk 8 above)
Comment (lanyue-llk) [reply moli-html2md/src/writer.rs]: 确认成立。原实现会把带尾部空格的相邻代码值变成两个空格，也会给原本无间隔的代码值新增空格。384d1d2a 在两个相邻 code 节点无法用纯 Markdown 无损分隔时，对前一节点使用内联 HTML；不新增可见字符，同时保留两个 code 节点。新增场景测试按实际 HTML 渲染结果检查完整文本和节点边界，覆盖默认和 preformatted 模式、空 code 与 Markdown 特殊字符；moli-html2md 全套测试通过。

**Review by lanyue-llk — COMMENTED** [review]
Deep re-review of `384d1d2a`: no additional blocking code findings.

I reran the earlier regression probes: empty leading blocks no longer invent a blank line, the source's own leading newline survives, empty/comment-only code nodes no longer inject backticks, inline output remains inline, and adjacent code nodes retain their original text without extra spaces. The previous findings are resolved on this revision.

I also tested the new inline-HTML path in 309 combinations covering both whitespace modes, ordinary text, strong/emphasis/strike, links, headings, lists, blockquotes and GFM tables. The probes check distinct code nodes and exact code text with backticks, backslashes, pipes, brackets, image/link-like syntax, HTML-looking text, entities, tabs and repeated/boundary spaces. All passed, including the table/pipe escaping interaction. The repository's unmodified `cargo test -p moli-html2md` suite passed all 294 tests.

The resulting design fits the documented structural-dump scope: disclosure retention lives in the shared strip path; preformatted boundaries use structural tags without layout dependencies; the existing inline-HTML approach handles code boundaries that cannot be represented losslessly by adjacent Markdown delimiters. No page-specific selectors or image-content heuristics remain in the final change.

Non-blocking description cleanup remains: update the Changes section to explain the lossless inline-HTML fallback, rather than saying every adjacent element becomes a Markdown code span, and condense the obsolete review/troubleshooting chronology into final behavior and validation. Keep the explicit statement that the 504-page measurement belongs to `30c4f85b`, not the current revision.

Current CI: Format, Clippy, Nextest, Python and Node checks have passed. Both release builds are still queued, so release-dependent end-to-end checks are not yet available for this head. This review does not represent those pending checks as successful.

File: moli-html2md/src/machine.rs [hunk 9]
```diff
@@ -435,17 +449,20 @@ fn class_language(class: Option<&str>) -> Option<&str> {
...
+fn is_structural_block(tag: &str) -> bool {
     matches!(
         tag,
         "address"
-            | "audio"
-            | "canvas"
-            | "frameset"
-            | "isindex"
-            | "noframes"
-            | "output"
```
Comment (lanyue-llk) [inline moli-html2md/src/machine.rs:446]: 这些删掉是确认过的吗？

File: moli-html2md/src/writer.rs [hunk 10]
```diff
@@ -446,6 +456,26 @@ impl<'a> Writer<'a> {
...
+
+    fn flush_code_html(&mut self) {
+        if let Some(code) = self.code.take() {
+            self.output.push_str("<code>");
+            for ch in code.chars() {
+                match ch {
+                    '&' => self.output.push_str("&amp;"),
+                    '<' => self.output.push_str("&lt;"),
+                    '>' => self.output.push_str("&gt;"),
+                    '\\' | '`' | '*' | '_' | '[' | ']' | '|' | '~' => {
```
Comment (lanyue-llk) [inline moli-html2md/src/writer.rs]: 这是不是穷举式的改法？是否会出现不在这里面的其它情况？是否有更好的方案？更通用，确定不是为当前问题打补丁

File: moli-html2md/tests/turndown/expectations.json [hunk 11]
```diff
@@ -4,8 +4,8 @@
     "reason": "Moli fences bare pre elements as well as pre/code, preserving preformatted text. Turndown only recognizes a pre/code pair."
   },
   "extra_adjacent_code": {
-    "markdown": "`ab`",
-    "reason": "Adjacent code elements contain ab. Turndown joins their delimiters into one code span whose text contains two extra backticks."
```
Comment (lanyue-llk) [inline moli-html2md/tests/turndown/expectations.json:8]: 新的能兼容原来的吗？确定删除吗

File: moli-protocol/src/domains/dom/tests/geometry.rs [hunk 12]
```diff
@@ -1388,9 +1388,9 @@ async fn scroll_into_view_if_needed_accepts_element_node() {
     let mut ctx = TestContext::new();
     load_bc(&mut ctx, "BID-A");
 
-    let rows = "<div>row</div>".repeat(60);
+    let spacer = "<div style='height:200vh'></div>";
```
Comment (lanyue-llk) [inline moli-protocol/src/domains/dom/tests/geometry.rs]: 新的修改兼容老的情况吗？需要兼容吗？

Comment (lanyue-llk) [reply moli-protocol/src/domains/dom/tests/geometry.rs]: 整个pr 都得考虑这个问题

File: moli-renderer-v8/src/runtime/page_dump.rs [hunk 13]
```diff
@@ -157,8 +157,6 @@ fn apply_strip_options(dom: &mut NativeDom, strip: RendererPageDumpStripOptions)
                         | "dialog"
                         | "menu"
                         | "menuitem"
-                        | "details"
-                        | "summary"
```
Comment (lanyue-llk) [inline moli-renderer-v8/src/runtime/page_dump.rs:161]: 确实整个pr 的删除和增加、修改，是向前兼容的，不是说不能增删，而是如果原来也是对的需求和功能，新修改后能保持，不能是删了原来的测试或者情况，测试通过了，就认为是修改对了，那不是正确的改法。

File: moli-html2md/src/machine.rs (hunk 9 above)
Comment (lanyue-llk) [reply moli-html2md/src/machine.rs:446]: 已核对完整 diff：这六个标签没有从原 is_block 行为中删除，而是保留在 is_block 的 OR 分支。新 is_structural_block 只给 pre 内部结构换行使用：output、canvas、audio 等内联内容不应强制换行；普通转换仍保留原来的判断。标题、列表、blockquote 在普通转换中已有专门分支。现有 output 内联样例与标题、列表换行样例同时约束两条路径。

File: moli-html2md/src/writer.rs (hunk 10 above)
Comment (lanyue-llk) [reply moli-html2md/src/writer.rs]: 维护性意见成立。4065e86c 把手写符号列表改为 ch.is_ascii_punctuation()，采用 CommonMark 定义的完整可反斜杠转义类别；HTML 的 &、<、> 仍用实体编码。新增场景包含命令、URL、Markdown 示例、实体、标点和 Unicode，用独立 HTML 解析器比较源 HTML 与最终渲染的 code 文本和节点，不断言转义实现字符串。未将列表外所有符号都声称为已确认缺陷。CI：https://github.com/lexmount/moli/actions/runs/35485830577 。规范：https://spec.commonmark.org/0.31.2/#backslash-escapes

File: moli-html2md/tests/turndown/expectations.json (hunk 11 above)
Comment (lanyue-llk) [reply moli-html2md/tests/turndown/expectations.json:8]: 没有删除原测试：extra_adjacent_code 仍使用原来的 <code>a</code><code>b</code> 输入。修改的是错误预期：原 `ab` 合并了两个 code 节点；新 <code>a</code>`b` 保留两个节点，合起来的文本仍是 ab，没有新增空格。单个 code、分隔文本和前后空白测试仍保留。新增语义回归检查最终渲染文本和节点。

File: moli-protocol/src/domains/dom/tests/geometry.rs (hunk 12 above)
Comment (lanyue-llk) [reply moli-protocol/src/domains/dom/tests/geometry.rs]: 覆盖面担忧成立。200vh 空占位保住了滚动前提，但替换了原来的文本行场景。4065e86c 恢复三个测试中的 60 个 row 文本 div，每行显式 height:4vh，保证目标在视口外并避免依赖本机字体行高。原 nodeId、objectId、display:contents 调用和 scrollY 断言保持，没有修改滚动生产代码。交由本次 CI 验证。

File: moli-renderer-v8/src/runtime/page_dump.rs (hunk 13 above)
Comment (lanyue-llk) [reply moli-renderer-v8/src/runtime/page_dump.rs:161]: 已逐项检查整个 PR。details/summary 无条件删除子树会丢失合法正文，因此纠正；nav、button 等原有清理规则保留。4065e86c 扩展同一端到端场景：关闭和打开的 details 正文保留，nav 内 details 和正文内按钮仍清除。同时恢复滚动测试的文本行场景，保留原 adjacent-code fixture，只纠正合并两个代码节点的错误预期。没有删除测试来规避旧需求；新提交 CI 正在执行，尚未标为通过。

**Comment by lanyue-llk** [comment]
本轮五个 inline 线程均已逐条回复，修复提交为 `4065e86c`。

CI 的 `cargo nextest run --workspace --profile ci` 已通过：18,286 passed，13 skipped；该配置没有测试名称过滤，本次新增语义测试及恢复文本行的滚动测试没有 ignore 标记。Clippy、Format、Node/Python harness 和 Windows storage 检查也已通过。release HEAD / common ancestor 构建仍在运行，未将整个 workflow 标为通过。

本轮实际改动是：采用完整 CommonMark ASCII 标点转义类别、恢复原文本行滚动场景并固定行高、增加清理 UI 与保留详情正文共存的场景，以及按源 HTML 与最终 Markdown 解析结果检查代码文本和节点。六个标签从原判断中删除、原 adjacent-code 输入测试被删除这两项疑虑不成立，具体依据已在对应线程说明。

[CI 与完整日志](https://github.com/lexmount/moli/actions/runs/35485830577)

**Review by ldm0 — COMMENTED** [review]
LGTM
